### ISSUE TITLE
Reddit provider validateAuthorizationCode not working as expected

### ISSUE THREAD

**Issue by dickwolff** [pilcrowonpaper/arctic#220]
I implemented Reddit using the Arctic Reddit provider, as I previously did with the likes of Twitter, Google. I have it configured like this:
```typescript
export const reddit = new Reddit(process.env.AUTH_REDDIT_ID!, process.env.AUTH_REDDIT_SECRET!, `${process.env.SITE_URL}api/auth/callback/reddit`);

…

async function createRedditUrl(state) {
    const url = await reddit.createAuthorizationURL(state, ["identity"]);
    url.searchParams.set("duration", "temporary");
    
    cookies().set("reddit_oauth_state", state, {
        path: "/",
        secure: process.env.NODE_ENV === "production",
        httpOnly: true,
        maxAge: 60 * 10,
        sameSite: "lax"
    });

    return url.href;
}
```

The Reddit callback however fails when it calls `validateAuthorizationCode(code)`. There is no detailed error, so I did not know what exactly failed. `code` is the result that comes back from the Reddit callback.

I then implemented the code myself using `fetch` and that worked.

```typescript
 const accessTokenResponse = await fetch("https://www.reddit.com/api/v1/access_token", {
     method: "POST",
     headers: {
         "Content-Type": "application/x-www-form-urlencoded",
         Authorization: `Basic ${btoa(process.env.AUTH_REDDIT_ID + ":" + process.env.AUTH_REDDIT_SECRET)}`,
     },
     body: new URLSearchParams({
         grant_type: "authorization_code",
         code: code,
         redirect_uri: `${process.env.SITE_URL}api/auth/callback/reddit`,
     }),
 });
```

On the surface my code looks the same as the code in Arctic’s `validateAuthorizationCode()` but somehow that doesn’t work.

**Comment by pilcrowonpaper** [comment]
What error does `validateAuthorizationCode()` throw? Is it a regular `Error` or one of Arctic's (`ArcticFetchError`, `OAuth2RequestError`)?

**Comment by dickwolff** [comment]
See the stacktrace below:

```
 error: {
    name: 'Error',
    message: 'Failed to parse response body',
    stack: 'Error: Failed to parse response body\n' +
      '    at sendTokenRequest (webpack-internal:///(rsc)/./node_modules/arctic/dist/request.js:41:15)\n' +
      '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n' +
      '    at async OAuth2Client.validateAuthorizationCode (webpack-internal:///(rsc)/./node_modules/arctic/dist/client.js:73:24)\n' +
      '    at async Reddit.validateAuthorizationCode (webpack-internal:///(rsc)/./node_modules/arctic/dist/providers/reddit.js:19:24)\n' +
```

**Comment by pilcrowonpaper** [comment]
Hm, that indicates the response body isn't valid JSON. You might need to manually add a `console.log()` in `sendTokenRequest()` to figure out what's actually being returned by Reddit.

**Comment by pilcrowonpaper** [comment]
Reading through the docs, maybe Reddit is returning a 401 response with no response body?

https://github.com/reddit-archive/reddit/wiki/oauth2

In which case, either your client ID or secret is invalid

**Comment by dickwolff** [comment]
When I send the command myself, it does work. So my environment variables are set. See the example I provided at the bottom of the issue. So I think it’s something within Arctic. 

I will try and debug and see if I can get a more detailed error message.

**Comment by joejemmely** [comment]
Some providers require the `client_id` to always be defined. I'm wondering if removing the if statement at https://github.com/pilcrowonpaper/arctic/blob/c843cb456726f429613961a8ae1ec01818bf3dd3/src/client.ts#L79-L81 would solve the issue.

**Comment by dickwolff** [comment]
So I dived into debugging this just now and firstly added some logging to get more details into the faulty response. This resulted in the following (which logs the `response` object in `sendTokenRequest()`: 

```js
    status: 411,
    .....
    statusText: 'Length Required',
```

For some reason when constructing the request with `createOAuth2Request()` where you create the fetch request and set the `URLSearchParams` object as body, when you then send the fetch request the `Content-Length` header is not set. When I log `request.headers` it gives me:

```js
HeadersList {
  cookies: null,
  [Symbol(headers map)]: Map(3) {
    'content-type' => {
      name: 'Content-Type',
      value: 'application/x-www-form-urlencoded'
    },
    'accept' => { name: 'Accept', value: 'application/json' },
    'user-agent' => { name: 'User-Agent', value: 'arctic' }
  },
  [Symbol(headers map sorted)]: null
}
```

So I went on looking for a fix, and put `request.headers.set("Content-Length", body.toString().length.toString());` inside `createOAuth2Request()`, below the current set of headers added there. After adding this line, the request goes through and I can log in using Reddit.

I don't know if this is the preferred way to fix this. I would assume the length would be set automatically when you construct a request. When you call the API using `fetch` with the example I provided in my initial issue description, it just works (without setting this header manually).
